libraries/AP_Avoidance/AP_Avoidance.h

```diff
@@ -0,0 +1,199 @@
+
+#pragma once
+
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  Situational awareness for ArduPilot
+
+ - record a series of moving points in space which should be avoided
+ - produce messages for GCS if a collision risk is detected
+
+  Peter Barker, May 2016
+
+  based on AP_ADSB,  Tom Pittenger, November 2015
+*/
+
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_ADSB/AP_ADSB.h>
+
+class AP_Avoidance {
+
+public:
+
+    class Sample {
+    public:
+        MAV_COLLISION_SRC src;
+        uint32_t src_id;
+        uint32_t timestamp_ms;
+
+        Location _location;
+        Vector3f _velocity;
+
+        // fields relating to this being a threat.  These would be the reason to have a separate list of threats:
+        MAV_COLLISION_THREAT_LEVEL threat_level;
+        float closest_approach_xy; // metres
+        float time_to_closest_approach; // seconds
+        float distance_to_closest_approach; // metres
+        uint32_t last_gcs_report_time; // millis
+    };
+
+    AP_Avoidance(AP_AHRS &ahrs, class AP_ADSB &adsb);
+
+    // for holding parameters
+    static const struct AP_Param::GroupInfo var_info[];
+
+    void add_sample(uint32_t sample_timestamp_ms,
+                    const MAV_COLLISION_SRC src,
+                    uint32_t src_id,
+                    const Location &loc,
+                    const Vector3f &vel); // vel is NED!
+
+    void add_sample(uint32_t sample_timestamp_ms,
+                    const MAV_COLLISION_SRC src,
+                    uint32_t src_id,
+                    const Location &loc,
+                    float cog,
+                    float hspeed,
+                    float vspeed);
+
+    void update();
+
+    static AP_Avoidance *instance(void) {
+        return _instance;
+    }
+
+    AP_Int8     _enabled;
+    AP_Int8     _samples_max;
+
+    AP_Int8     _action_f;
+    AP_Int8     _action_w;
+
+    AP_Int8     _recovery_f;
+    AP_Int8     _recovery_w;
+
+    AP_Int8     _time_horizon_warn_xy;
+    AP_Int8     _time_horizon_warn_z;
+    AP_Int16	_distance_warn_xy;
+    AP_Int16	_distance_warn_z;
+
+    AP_Int8     _time_horizon_fail_xy;
+    AP_Int8     _time_horizon_fail_z;
+    AP_Int16	_distance_fail_xy;
+    AP_Int16	_distance_fail_z;
+
+protected:
+
+    enum state_t {
+        STATE_CLEAR = 0,
+        STATE_WARN = 1,
+        STATE_FAIL = 2,
```

Comment:
I've standardised on e.g. _fail_distance_xy and _warn_distance_xy.

My reasoning "fail" vs "failsafe" is twofold: firstly there's no "warnsafe".  Secondly, it's the entire mechanism which is the failsafe, not just the action taken when a collision is imminent.